cli/src/commands/gerrit/upload.rs

```diff
@@ -0,0 +1,389 @@
+            description.trim(),
+            if trailers.is_empty() { "\n\n" } else { "\n" },
+            gerrit_change_id
+        );
+
+        let new_parents = original_commit
+            .parents()
+            .map(|parent| {
+                let p = parent.unwrap();
+                if let Some(rewritten_parent) = old_to_new.get(p.id()) {
+                    rewritten_parent
+                } else {
+                    &p
+                }
+                .id()
+                .clone()
+            })
+            .collect();
+
+        // rewrite the set of parents to point to the commits that were
+        // previously rewritten in toposort order
+        //
+        // TODO FIXME (aseipp): this whole dance with toposorting, calculating
+        // new_parents, and then doing rewrite_commit is roughly equivalent to
+        // what we do in duplicate.rs as well. we should probably refactor this?
+        let new_commit = tx
+            .repo_mut()
+            .rewrite_commit(&original_commit)
+            .set_description(new_description)
+            .set_parents(new_parents)
+            // Set the timestamp back to the timestamp of the original commit.
+            // Otherwise, `jj gerrit upload @ && jj gerrit upload @` will upload
+            // two patchsets with the only difference being the timestamp.
+            .set_committer(original_commit.committer().clone())
+            .set_author(original_commit.author().clone())
+            .write()?;
+
+        old_to_new.insert(original_commit.id().clone(), new_commit.clone());
+    }
+    writeln!(ui.stderr())?;
+
+    let remote_ref = format!("refs/for/{remote_branch}");
+    writeln!(
+        ui.stderr(),
+        "Found {} heads to push to Gerrit (remote '{}'), target branch '{}'",
+        old_heads.len(),
+        remote,
+        remote_branch,
+    )?;
+
+    writeln!(ui.stderr())?;
+
+    // NOTE (aseipp): because we are pushing everything to the same remote ref,
+    // we have to loop and push each commit one at a time, even though
+    // push_updates in theory supports multiple GitRefUpdates at once, because
+    // we obviously can't push multiple heads to the same ref.
+    for head in &old_heads {
+        write!(
+            ui.stderr(),
+            "{}",
+            if upload.dry_run {
+                "Dry-run: Would push "
+            } else {
+                "Pushing "
+            }
+        )?;
+        // We have to write the old commit here, because the until we finish
+        // the transaction (which we don't), the new commit is labelled as
+        // "hidden".
+        tx.base_workspace_helper().write_commit_summary(
+            ui.stderr_formatter().as_mut(),
+            &store.get_commit(head).unwrap(),
+        )?;
+        writeln!(ui.stderr())?;
+
+        if upload.dry_run {
+            continue;
+        }
+
+        let new_commit = store
+            .get_commit(old_to_new.get(head).unwrap().id())
+            .unwrap();
```

Comment:
similar to avoid, this seems to extract the id from a commit only to look up the commit again...